IIM-ILE examples

AminKolah opened the pull request:
<!--
This template should be included in all pull requests. Items in the list should
either be completed by the original author or explicitly dismissed by one of the
IBAMR principal developers.

IBAMR is a community effort and it wouldn't exist without people contributing
code. Thanks in advance for helping to make IBAMR better!
-->

### IBAMR Pull Request Checklist
- [ ] Does the test suite pass?
- [ ] Was clang-format run?
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [ ] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

AminKolah commented:
I have autoconf 2.71 installed on my macbook but when I run bootstrap I get an error that seems to be saying my autoconf is old but I don't get why it says that). Any idea? Also below is the rest of the errors I get

configure.ac:18: error: Autoconf version 2.71 or higher is required
configure.ac:18: the top level
autom4te: /usr/bin/m4 failed with exit status: 63
aclocal: error: echo failed with exit status: 63
autoreconf: aclocal failed with exit status: 63
This script encountered an unhandled error when trying to install autotools.

drwells commented:
In addition - it looks like you added a bunch of log files and other temporary output. I recommend using a build directory so that those files don't get added to the base directory.

AminKolah commented:
Okay, I guess from two hundred something files to 85 files is a progress!

AminKolah commented:
@drwells any ideas why other CI tests are still failing? I'm not sure if I totally understand it from the details provided for the errors.

drwells commented:
Here's the relevant part of the log:
```
2022-06-15T16:02:31.2454895Z CMake Error: File /__w/IBAMR/IBAMR/examples/IIM/ex1/input3d does not exist.
2022-06-15T16:02:31.2455171Z CMake Error at examples/CMakeLists.txt:90 (CONFIGURE_FILE):
2022-06-15T16:02:31.2455580Z   CONFIGURE_FILE Problem configuring file
2022-06-15T16:02:31.2455769Z Call Stack (most recent call first):
2022-06-15T16:02:31.2455975Z   examples/IIM/ex1/CMakeLists.txt:35 (IBAMR_ADD_EXAMPLE)
2022-06-15T16:02:31.2456113Z 
2022-06-15T16:02:31.2456120Z 
2022-06-15T16:02:31.2463550Z CMake Error: File /__w/IBAMR/IBAMR/examples/IIM/ex1/airfoil_bc_0.1_TRI3_0.1.e does not exist.
2022-06-15T16:02:31.2463915Z CMake Error at examples/CMakeLists.txt:94 (CONFIGURE_FILE):
2022-06-15T16:02:31.2464195Z   CONFIGURE_FILE Problem configuring file
2022-06-15T16:02:31.2464448Z Call Stack (most recent call first):
2022-06-15T16:02:31.2464868Z   examples/IIM/ex1/CMakeLists.txt:35 (IBAMR_ADD_EXAMPLE)
2022-06-15T16:02:31.2465047Z 
2022-06-15T16:02:31.2465053Z 
2022-06-15T16:02:31.2465225Z CMake Error: File /__w/IBAMR/IBAMR/examples/IIM/ex2/input3d.iim does not exist.
2022-06-15T16:02:31.2465559Z CMake Error at examples/CMakeLists.txt:90 (CONFIGURE_FILE):
2022-06-15T16:02:31.2465840Z   CONFIGURE_FILE Problem configuring file
2022-06-15T16:02:31.2466087Z Call Stack (most recent call first):
2022-06-15T16:02:31.2466363Z   examples/IIM/ex2/CMakeLists.txt:15 (IBAMR_ADD_EXAMPLE)
2022-06-15T16:02:31.2466528Z 
2022-06-15T16:02:31.2470312Z 
2022-06-15T16:02:31.2481460Z CMake Error: File /__w/IBAMR/IBAMR/examples/IIM/ex4/compute_St.m does not exist.
2022-06-15T16:02:31.2481804Z CMake Error at examples/CMakeLists.txt:94 (CONFIGURE_FILE):
2022-06-15T16:02:31.2482089Z   CONFIGURE_FILE Problem configuring file
2022-06-15T16:02:31.2482337Z Call Stack (most recent call first):
2022-06-15T16:02:31.2482609Z   examples/IIM/ex4/CMakeLists.txt:14 (IBAMR_ADD_EXAMPLE)
2022-06-15T16:02:31.2482769Z 
2022-06-15T16:02:31.2482783Z 
2022-06-15T16:02:31.2503592Z CMake Warning (dev) at examples/IIM/ex7/CMakeLists.txt:19:
2022-06-15T16:02:31.2503810Z   Syntax Warning in cmake code at column 43
2022-06-15T16:02:31.2503921Z 
2022-06-15T16:02:31.2504031Z   Argument not separated from preceding token by whitespace.
2022-06-15T16:02:31.2504495Z This warning is for project developers.  Use -Wno-dev to suppress it.
2022-06-15T16:02:31.2504646Z 
2022-06-15T16:02:31.2516914Z CMake Error: File /__w/IBAMR/IBAMR/examples/IIM/ex8/damped_ds_0.1_QUAD9_lower_AllSides.e does not exist.
2022-06-15T16:02:31.2517464Z CMake Error at examples/CMakeLists.txt:94 (CONFIGURE_FILE):
2022-06-15T16:02:31.2517752Z   CONFIGURE_FILE Problem configuring file
2022-06-15T16:02:31.2517995Z Call Stack (most recent call first):
2022-06-15T16:02:31.2518269Z   examples/IIM/ex8/CMakeLists.txt:15 (IBAMR_ADD_EXAMPLE)
2022-06-15T16:02:31.2518436Z 
2022-06-15T16:02:31.2518442Z 
2022-06-15T16:02:31.2526693Z CMake Error: File /__w/IBAMR/IBAMR/examples/IIM/ex8/damped_ds_0.1_QUAD9_upper_AllSides.e does not exist.
2022-06-15T16:02:31.2527114Z CMake Error at examples/CMakeLists.txt:94 (CONFIGURE_FILE):
2022-06-15T16:02:31.2527405Z   CONFIGURE_FILE Problem configuring file
2022-06-15T16:02:31.2527658Z Call Stack (most recent call first):
2022-06-15T16:02:31.2527921Z   examples/IIM/ex8/CMakeLists.txt:15 (IBAMR_ADD_EXAMPLE)
2022-06-15T16:02:31.2528097Z 
2022-06-15T16:02:31.2528103Z 
2022-06-15T16:02:31.2528416Z CMake Error: File /__w/IBAMR/IBAMR/examples/IIM/ex9/Hessenthaler_thin_mesh.e does not exist.
2022-06-15T16:02:31.2528775Z CMake Error at examples/CMakeLists.txt:94 (CONFIGURE_FILE):
2022-06-15T16:02:31.2529057Z   CONFIGURE_FILE Problem configuring file
2022-06-15T16:02:31.2529304Z Call Stack (most recent call first):
2022-06-15T16:02:31.2529576Z   examples/IIM/ex9/CMakeLists.txt:15 (IBAMR_ADD_EXAMPLE)
2022-06-15T16:02:31.2529742Z 
2022-06-15T16:02:31.2529747Z 
2022-06-15T16:02:31.3128213Z -- Configuring incomplete, errors occurred!
2022-06-15T16:02:31.3128555Z See also "/build/ibamr/CMakeFiles/CMakeOutput.log".
2022-06-15T16:02:31.3128866Z See also "/build/ibamr/CMakeFiles/CMakeError.log".
2022-06-15T16:02:31.3279117Z Error(s) when configuring the project
2022-06-15T16:02:31.3279423Z Command exited with the value: 1
```

I think the extra files you listed in the `CMakeLists.txt` files are out of sync with the ones you put in the PR.

AminKolah commented:
> 

Thanks!

drwells commented:
This looks good - we can cleanly compile the examples with the oldest and second most recent libMesh without any warnings. I'll double check the examples this evening.

AminKolah commented:
That's great.. I was going to write I sloppily/lazily introduced a bunch of static variables in most examples for tether parameters but I think we would want to put them into a tether data structure and pass them via input file as ctx data pointer into tether functions...I can set this up and finish them in the next couple of days, if you think it's necessary

drwells commented on examples/IIM/ex9/BcData.h:
```diff
@@ -0,0 +1,36 @@
+    const double z_min;
+    const double z_max;
+    
+        BcData(Pointer<Database> input_db)
```
We don't do a good job enforcing it but all our headers should compile on their own - i.e., we need the headers for `tbox::Pointer` and `tbox::Database` here.

drwells commented on examples/IIM/ex2/example.cpp:
```diff
@@ -0,0 +1,1237 @@
+                    {
+                       // const double rsrc = d_rsrc[side];
+						//
+                        Vector n;
```
```suggestion
                        VectorNd n;
```
This is a good example of something which fails to compile with newer versions of Eigen since we used `app_namespaces.h`.

drwells commented on examples/IIM/ex2/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+  EXAMPLE_GROUP
+    examples-IIM
+  SOURCES
+    example.cpp
```
```suggestion
    example.cpp
    FeedbackForcer.cpp
```

drwells reviewed (changes requested):
We could use a neater style than global static variables but it is what we do in the other tutorials and it is sufficiently simple for small things.

One thing I noticed is that we are not actually checking the example compilation with warnings on and CMake - I'll go fix that.

drwells commented:
I updated the CI to compile the examples with more configurations and we now get a bunch of easy-to-fix warnings, e.g.,
```cpp

/__w/IBAMR/IBAMR/examples/IIM/ex9/example.cpp: In function ‘void ModelData::tether_FSI_force_function_housing(libMesh::VectorValue<double>&, const libMesh::VectorValue<double>&, const libMesh::VectorValue<double>&, const libMesh::TensorValue<double>&, const libMesh::Point&, const libMesh::Point&, libMesh::Elem*, short unsigned int, const std::vector<const std::vector<double>*>&, const std::vector<const std::vector<libMesh::VectorValue<double> >*>&, double, void*)’:
/__w/IBAMR/IBAMR/examples/IIM/ex9/example.cpp:160:35: error: unused parameter ‘elem’ [-Werror=unused-parameter]
  160 |                       Elem* const elem,
```
Can you fix those and then rebase? Let me know if you need a hand :)

drwells commented:
Looks good - I'll double-check the examples a little later this morning.

Can you rebase and squash your commits?

AminKolah commented:
Thanks David. I squashed a couple of times at least. I also rebased at some point (can you check if changes related to CMake are in the commit history?).  Let me know if you want me to squash more commits (though it seems I won't be able to squash a bunch of commits when they are not leading up to HEAD).

drwells commented on examples/IIM/ex6/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+  REQUIRES
+    IBAMR_HAVE_LIBMESH
+  LINK_TARGETS
+    IBAMR3d
```
```suggestion
    IBAMR2d
```

drwells commented on examples/IIM/ex3/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+  INPUT_FILES
+    input2d
+  EXTRA_FILES
+    compute_St.m
```
```suggestion
    compute_St.m
    airfoil_bc_0.1_TRI3_0.1.e
```

AminKolah commented:
To perform the rebase, I first fetched upstream in my main IBAMR fork and then rebased this branch by drag-and-dropping it onto master via GitKraken. Wasn't that right? ..Seems like I always wind with a messed up commit history with rebase

drwells commented:
I haven't used git-kraken so I don't know for sure, but I think that does a merge and not a rebase - 2f41181dbad3e1f546e21c9d6608aa111cff51ad is definitely a merge commit.

We definitely do not want a merge since, as master continues to change, the history becomes tangled up if there are multiple merges. In contrast, a rebase puts all your commits together on top of `master` which is what we always want.

at the command line rebasing is pretty easy (and surely git-kraken has some option for this too):
```sh
> git rebase master
> # go and fix and conflicts
> git rebase --continue
> # etc
```
